appserver/payara-appserver-modules/microprofile/fault-tolerance/src/test/java/fish/payara/microprofile/faulttolerance/policy/AbstractBulkheadTest.java

```diff
@@ -134,7 +134,7 @@ public void delay(long delayMillis) throws InterruptedException {
     static class InOut {
         static final Object IN = new Object();
 
-        Object result;
+        volatile Object result;
```

Comment:
As discussed I believe the issue here is that the async worker entering and exiting the bulkhead also sets the result in test code after having left the bulkhead. At that point the main thread can see all workers exited the bulkhead and starts to assert but there is no synchronisation forcing it to wait until the worker also has set its result. This part has to be outside of the bulkhead. So a proper fix is to introduce some kind of waiting in the main thread that keeps track of the workers and if they have set the result. This is by no means a request to do such a change in this PR but meant to document the issue. We should address it in a separate PR.